controller/utils/fio-chain.js

```diff
@@ -1,27 +1,101 @@
 import 'dotenv/config';
 
-import fetch from 'node-fetch';
+import { Fio } from '@fioprotocol/fiojs';
+import * as textEncoderObj from 'text-encoding';
 
 import config from '../../config/config.js';
-import { FIO_ACCOUNT_NAMES } from '../constants/chain.js';
-import { MINUTE_IN_MILLISECONDS } from '../constants/general.js';
-import {
-  checkHttpResponseStatus,
-  sleep,
-  rateLimiterFor1000Rpm,
-} from '../utils/general.js';
+import { handleChainError } from '../../controller/utils/log-files.js';
+import { FIO_ACCOUNT_NAMES, FIO_TABLE_NAMES } from '../constants/chain.js';
+import { checkHttpResponseStatus, fetchWithRateLimit } from '../utils/general.js';
+
+const defaultTextEncoderObj = textEncoderObj.default || {};
+
+const TextDecoder = defaultTextEncoderObj.TextDecoder;
+const TextEncoder = defaultTextEncoderObj.TextEncoder;
+
+const textDecoder = new TextDecoder();
+const textEncoder = new TextEncoder();
 
 const {
   fio: {
-    FIO_SERVER_HISTORY_VERSION,
     FIO_SERVER_URL_HISTORY,
-    FIO_SERVER_URL_ACTION,
     FIO_SERVER_URL_HISTORY_BACKUP,
+    FIO_SERVER_URL_ACTION,
+    FIO_SERVER_URL_ACTION_BACKUP,
+    FIO_GET_TABLE_ROWS_OFFSET,
+    FIO_ORACLE_PERMISSION,
+    FIO_ORACLE_PRIVATE_KEY,
+    FIO_ORACLE_ACCOUNT,
   },
 } = config;
 
+const makeGetInfoUrl = (baseUrl) => `${baseUrl}v1/chain/get_info`;
+const makeGetBlockUrl = (baseUrl) => `${baseUrl}v1/chain/get_block`;
+const makeGetRawAbiUrl = (baseUrl) => `${baseUrl}v1/chain/get_raw_abi`;
+const makePushTransactionUrl = (baseUrl) => `${baseUrl}v1/chain/push_transaction`;
+const makeTableRowsUrl = (baseUrl) => `${baseUrl}v1/chain/get_table_rows`;
+const makeDeltasUrl = ({ baseUrl, params }) => {
+  const queryString = new URLSearchParams(params).toString();
+  return `${baseUrl}v2/history/get_deltas?${queryString}`;
+};
+
+const getFioChainInfo = async () =>
+  await (
+    await fetchWithRateLimit({
+      url: makeGetInfoUrl(FIO_SERVER_URL_ACTION),
+      backupUrl: FIO_SERVER_URL_ACTION_BACKUP
+        ? makeGetInfoUrl(FIO_SERVER_URL_ACTION_BACKUP)
+        : null,
+    })
+  ).json();
+
+const getFioBlockInfo = async (lastIrreversibleBlock) =>
+  await (
+    await fetchWithRateLimit({
+      url: makeGetBlockUrl(FIO_SERVER_URL_ACTION),
+      options: {
+        body: `{"block_num_or_id": ${lastIrreversibleBlock}}`,
```

Comment:
would it better to use `JSON.stringify({block_num_or_id: lastIrreversibleBlock})`

Reply:
Done